File: ios/NmpGallery/NmpGallery/EmbedCard.swift
import SwiftUI

/// Embedded event card (quoted note / nevent / naddr). Applies the Swift
/// `RenderContext` mirror: collapses at depth >= 4 (PD-015) or on a
/// `visited`-set cycle, and degrades gracefully for dangling / unsupported
/// targets (D1 — never blank, never a spinner).
struct EmbedCard: View {
    let uri: String
    let refID: String
    let entry: EmbedEntry?
    let embeds: [String: EmbedEntry]
    let ctx: RenderContext

    var body: some View {
        guard let entry else {
            return AnyView(stub("Quoted event unavailable",
                                detail: refID, system: "questionmark.app"))
        }

        // Bundle-time context-independent collapse facts.
        if entry.collapsed {
            switch entry.collapseReason {
            case "dangling":
                return AnyView(stub("Quoted event unavailable",
                                    detail: refID,
                                    system: "wifi.slash"))
            case "unsupported":
                return AnyView(stub(
                    "Unsupported event (kind \(entry.resolvedKind))",
                    detail: refID, system: "shippingbox"))
            default:
                return AnyView(stub("Embed collapsed",
                                    detail: refID, system: "rectangle.slash"))
            }
        }

        // Render-time PD-015 depth + cycle guard (Swift RenderContext).
        if let ev = entry.event {
            let key = visitedKey(for: ev)
            let (collapse, reason) = ctx.shouldCollapse(into: key)
            if collapse {
                let label = reason == "cycle"
                    ? "Already shown (cycle broken)"
                    : "Quoted event (tap to open)"
                return AnyView(stub(label, detail: refID,
                                    system: reason == "cycle"
                                        ? "arrow.triangle.2.circlepath"
                                        : "chevron.right.circle"))
            }
        }

        return AnyView(card(entry))
    }

    @ViewBuilder
    private func card(_ entry: EmbedEntry) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let article = entry.article {
                ArticlePreview(header: article)
            } else if let list = entry.list {
                ListCard(list: list)
            }
            if let ev = entry.event {
                HStack(spacing: 6) {
                    Identicon(seed: ev.pubkey)
                        .frame(width: 18, height: 18)
                    Text("kind \(ev.kind) · @npub1\(ev.pubkey.prefix(6))…")
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
            if let body = entry.rendered, entry.article == nil,
               entry.list == nil {
                SegmentDtoView(
                    tree: body, embeds: embeds,
                    ctx: childCtx(entry))
            }
        }
        .padding(10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color.secondary.opacity(0.4), lineWidth: 1))
        .background(Color(.tertiarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }

    private func childCtx(_ entry: EmbedEntry) -> RenderContext {
        guard let ev = entry.event else { return ctx }
        return ctx.descend(into: visitedKey(for: ev))
    }

    @ViewBuilder
    private func stub(_ title: String, detail: String,
                      system: String) -> some View {
        HStack(spacing: 8) {
            Image(systemName: system)
                .foregroundStyle(.secondary)
            VStack(alignment: .leading, spacing: 2) {
                Text(title).font(.caption.bold())
                Text(detail.prefix(28) + "…")
                    .font(.caption2.monospaced())
                    .foregroundStyle(.tertiary)
            }
        }
        .padding(8)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color.secondary.opacity(0.10))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }
}

/// Medium-like article preview card (naddr → kind:30023).
struct ArticlePreview: View {
    let header: ArticleHeaderDto

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Label("Article", systemImage: "doc.richtext")
                .font(.caption2.bold())
                .foregroundStyle(.teal)
            Text(header.title ?? "(untitled)")
                .font(.headline)
            if let s = header.summary {
                Text(s).font(.caption).foregroundStyle(.secondary)
            }
            HStack(spacing: 6) {
                Identicon(seed: header.author)
                    .frame(width: 16, height: 16)
                Text("@npub1\(header.author.prefix(6))…")
                    .font(.caption2).foregroundStyle(.secondary)
                Spacer()
                Text("Read article ›")
                    .font(.caption2.bold())
                    .foregroundStyle(.teal)
            }
        }
        .padding(8)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color.teal.opacity(0.10))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }
}

/// NIP-51 inline titled list card (follow set / bookmarks / relay list).
struct ListCard: View {
    let list: ListDto

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Label(list.title ?? "List", systemImage: "list.bullet")
                .font(.caption.bold())
                .foregroundStyle(.green)
            if list.rows.isEmpty {
                Text("(no members)")
                    .font(.caption2).italic()
                    .foregroundStyle(.tertiary)
            }
            ForEach(Array(list.rows.enumerated()), id: \.offset) {
                _, row in
                rowView(row)
            }
        }
        .padding(8)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color.green.opacity(0.10))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }

    @ViewBuilder
    private func rowView(_ row: ListRowDto) -> some View {
        switch row {
        case let .profile(pubkey, name, _):
            HStack(spacing: 6) {
                Identicon(seed: pubkey).frame(width: 16, height: 16)
                Text(name.map { "@\($0)" }
                    ?? "@npub1\(pubkey.prefix(6))…")
                    .font(.caption)
            }
        case let .event(id):
            Label("note \(id.prefix(10))…", systemImage: "text.quote")
                .font(.caption2)
        case let .address(coord):
            Label(coord, systemImage: "doc.text")
                .font(.caption2).lineLimit(1)
        case let .hashtag(tag):
            Label("#\(tag)", systemImage: "number")
                .font(.caption2)
        case let .relay(url, read, write):
            HStack(spacing: 4) {
                Image(systemName: "antenna.radiowaves.left.and.right")
                Text(url).font(.caption2.monospaced())
                if read { tagBadge("R") }
                if write { tagBadge("W") }
            }
        case let .unknown(t):
            Text("[\(t)]").font(.caption2).foregroundStyle(.red)
        }
    }

    @ViewBuilder
    private func tagBadge(_ s: String) -> some View {
        Text(s)
            .font(.caption2.bold())
            .padding(.horizontal, 4)
            .background(Color.green.opacity(0.25))
            .clipShape(Capsule())
    }
}

File: android/gallery/src/main/java/org/nmp/gallery/ui/EmbedCard.kt
package org.nmp.gallery.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import org.nmp.gallery.model.ArticleHeaderDto
import org.nmp.gallery.model.EmbedEntry
import org.nmp.gallery.model.ListDto
import org.nmp.gallery.model.ListRowDto
import org.nmp.gallery.model.RenderContext
import org.nmp.gallery.model.visitedKey

/**
 * Compose port of Swift `EmbedCard`. Embedded event card (quoted note /
 * nevent / naddr). Applies the [RenderContext] PD-015 depth + cycle guard
 * and degrades gracefully for dangling / unsupported targets (D1 — never
 * blank, never a spinner).
 */
@Composable
fun EmbedCard(
    uri: String,
    refId: String,
    entry: EmbedEntry?,
    embeds: Map<String, EmbedEntry>,
    ctx: RenderContext,
) {
    if (entry == null) {
        StubBox("Quoted event unavailable", refId, glyph = "?")
        return
    }

    // Bundle-time, context-independent collapse facts.
    if (entry.collapsed) {
        when (entry.collapseReason) {
            "dangling" -> StubBox("Quoted event unavailable", refId, glyph = "⌀")
            "unsupported" -> StubBox(
                "Unsupported event (kind ${entry.resolvedKind})",
                refId,
                glyph = "▢",
            )
            else -> StubBox("Embed collapsed", refId, glyph = "▷")
        }
        return
    }

    // Render-time PD-015 depth + cycle guard.
    val ev = entry.event
    if (ev != null) {
        val key = visitedKey(ev)
        val (collapse, reason) = ctx.shouldCollapse(key)
        if (collapse) {
            val label = if (reason == "cycle") {
                "Already shown (cycle broken)"
            } else {
                "Quoted event (tap to open)"
            }
            val glyph = if (reason == "cycle") "↻" else "›"
            StubBox(label, refId, glyph = glyph)
            return
        }
    }

    CardBody(entry, embeds, ctx)
}

@Composable
private fun CardBody(
    entry: EmbedEntry,
    embeds: Map<String, EmbedEntry>,
    ctx: RenderContext,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(Color.Gray.copy(alpha = 0.06f))
            .border(
                width = 1.dp,
                color = MaterialTheme.colorScheme.outline.copy(alpha = 0.4f),
                shape = RoundedCornerShape(8.dp),
            )
            .padding(10.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        entry.article?.let { ArticlePreview(it) }
        entry.list?.let { ListCard(it) }

        entry.event?.let { ev ->
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Identicon(ev.pubkey, modifier = Modifier.size(18.dp))
                Text(
                    "kind ${ev.kind} · @npub1${ev.pubkey.take(6)}…",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        val body = entry.rendered
        if (body != null && entry.article == null && entry.list == null) {
            val childCtx = entry.event?.let { ctx.descend(visitedKey(it)) } ?: ctx
            SegmentDtoView(tree = body, embeds = embeds, ctx = childCtx)
        }
    }
}

@Composable
private fun StubBox(title: String, detail: String, glyph: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(Color.Gray.copy(alpha = 0.10f))
            .padding(8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(
            glyph,
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Column {
            Text(
                title,
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.Bold,
            )
            Text(
                detail.take(28) + "…",
                style = MaterialTheme.typography.labelSmall.copy(
                    fontFamily = FontFamily.Monospace,
                ),
                color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f),
            )
        }
    }
}

/** Medium-like article preview card (naddr → kind:30023). */
@Composable
fun ArticlePreview(header: ArticleHeaderDto) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(SwiftTeal.copy(alpha = 0.10f))
            .padding(8.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            "[Article]",
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Bold,
            color = SwiftTeal,
        )
        Text(
            header.title ?: "(untitled)",
            style = MaterialTheme.typography.titleSmall,
        )
        header.summary?.let { s ->
            Text(
                s,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            modifier = Modifier.fillMaxWidth(),
        ) {
            Identicon(header.author, modifier = Modifier.size(16.dp))
            Text(
                "@npub1${header.author.take(6)}…",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(Modifier.weight(1f))
            Text(
                "Read article ›",
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.Bold,
                color = SwiftTeal,
            )
        }
    }
}

/** NIP-51 inline titled list card (follow set / bookmarks / relay list). */
@Composable
fun ListCard(list: ListDto) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(SwiftGreen.copy(alpha = 0.10f))
            .padding(8.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            "▤ ${list.title ?: "List"}",
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.Bold,
            color = SwiftGreen,
        )
        if (list.rows.isEmpty()) {
            Text(
                "(no members)",
                style = MaterialTheme.typography.labelSmall.copy(
                    fontStyle = FontStyle.Italic,
                ),
                color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f),
            )
        }
        for (row in list.rows) ListRow(row)
    }
}

@Composable
private fun ListRow(row: ListRowDto) {
    when (row) {
        is ListRowDto.Profile -> Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Identicon(row.pubkey, modifier = Modifier.size(16.dp))
            Text(
                row.name?.let { "@$it" } ?: "@npub1${row.pubkey.take(6)}…",
                style = MaterialTheme.typography.labelMedium,
            )
        }
        is ListRowDto.Event -> Text(
            "“ note ${row.id.take(10)}…",
            style = MaterialTheme.typography.labelSmall,
        )
        is ListRowDto.Address -> Text(
            "▤ ${row.coord}",
            style = MaterialTheme.typography.labelSmall,
            maxLines = 1,
        )
        is ListRowDto.Hashtag -> Text(
            "# ${row.tag}",
            style = MaterialTheme.typography.labelSmall,
        )
        is ListRowDto.Relay -> Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text("📡", style = MaterialTheme.typography.labelSmall)
            Text(
                row.url,
                style = MaterialTheme.typography.labelSmall.copy(
                    fontFamily = FontFamily.Monospace,
                ),
            )
            if (row.read) TagBadge("R")
            if (row.write) TagBadge("W")
        }
        is ListRowDto.Unknown -> Text(
            "[${row.type}]",
            style = MaterialTheme.typography.labelSmall,
            color = SwiftRed,
        )
    }
}

@Composable
private fun TagBadge(label: String) {
    Text(
        label,
        style = MaterialTheme.typography.labelSmall,
        fontWeight = FontWeight.Bold,
        modifier = Modifier
            .clip(RoundedCornerShape(percent = 50))
            .background(SwiftGreen.copy(alpha = 0.25f))
            .padding(horizontal = 4.dp),
    )
}
